# Exercise 4: ANN for Multiclass Classification

# Objective : Learning the construction and training of ANN for Multi-class classfication problem

# Problem : 

# (1)  Design ANN with 3 hidden layers with 8,16,8 neurons  and appropriate input and output layers for Egyptian Skull data
# (2)  Obtain various classification metrics of your choice based on the training data
# (3)  Illustrate how out of sample prediction can be made


# Data Preparation

In [1]:
from google.colab import files
uploaded=files.upload()

Saving egypt.csv to egypt.csv


In [2]:
import pandas as pd
import io
df=pd.read_csv(io.BytesIO(uploaded['egypt.csv']))
df.head()

MB   BH   BL  NH  Year
0  131  138   89  49 -4000
1  125  131   92  48 -4000
2  131  132   99  50 -4000
3  119  132   96  44 -4000
4  136  143  100  54 -4000

# Seperation of features columns

In [3]:
X=df.iloc[:,:4]
X.head()

MB   BH   BL  NH
0  131  138   89  49
1  125  131   92  48
2  131  132   99  50
3  119  132   96  44
4  136  143  100  54

# Encoding the last column values using numbers 0,1,2,3,4

In [4]:
df2=df.replace(to_replace=[-4000,-3300,-1850,-200,150],value=[0,1,2,3,4])

# Creation of Target column and the encoded matrix

In [5]:
y=df2.iloc[:,4:5]

In [6]:
from keras.utils import to_categorical
y=to_categorical(y,num_classes=5)


# Building the network

In [7]:
from keras import models, layers
model=models.Sequential()
model.add(layers.Dense(32,activation='relu',input_dim=4)) # hidden layer - 1 with 32 neurons receiving 4 inputs
model.add(layers.Dense(16,activation='elu')) # hidden layer - 2 with 16 neurons
model.add(layers.Dense(16,activation='elu')) # hidden layer - 2 with 16 neurons
model.add(layers.Dense(8,activation='elu')) # hidden layer - 3 with 8 neurons
model.add(layers.Dense(5,activation='softmax'))
print(model.summary()) 

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 32)                160       
                                                                 
 dense_1 (Dense)             (None, 16)                528       
                                                                 
 dense_2 (Dense)             (None, 16)                272       
                                                                 
 dense_3 (Dense)             (None, 8)                 136       
                                                                 
 dense_4 (Dense)             (None, 5)                 45        
                                                                 
Total params: 1,141
Trainable params: 1,141
Non-trainable params: 0
_________________________________________________________________
None


# Compilation

In [8]:
model.compile(loss='categorical_crossentropy',optimizer='adam', metrics=['accuracy'])

# Fitting the model

In [ ]:
hist=model.fit(X,y,epochs=1000)

# Inline Prediction

In [ ]:
predicted=model.predict(X)
predicted

In [11]:
import numpy as np
y_pred=np.argmax(predicted,axis=1)
y_pred

array([2, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 3, 0, 1, 1, 1, 1, 3, 2, 1, 1, 1,
       3, 1, 1, 3, 3, 1, 0, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 3, 1, 1, 1, 3, 1, 3, 1, 3, 3, 1, 3, 1, 2, 1, 1, 1,
       2, 3, 2, 3, 1, 2, 1, 3, 1, 1, 3, 4, 3, 3, 3, 1, 3, 1, 3, 3, 1, 1,
       3, 4, 1, 3, 4, 1, 4, 3, 3, 4, 3, 1, 3, 3, 1, 3, 3, 3, 1, 1, 3, 3,
       3, 4, 3, 1, 3, 3, 3, 3, 3, 1, 4, 2, 3, 3, 4, 1, 3, 1, 1, 3, 4, 3,
       3, 3, 3, 2, 1, 3, 3, 4, 2, 1, 4, 3, 1, 4, 4, 1, 4, 1])

# Various classification metrics of your choice based on the training data

In [18]:
y_true=np.argmax(y,axis=1) 
from sklearn.metrics import precision_score 
from sklearn.metrics import recall_score 
from sklearn.metrics import f1_score
pr=precision_score(y_pred,y_true,average='weighted') 
re=recall_score(y_pred,y_true,average='weighted') 
f1=f1_score(y_pred,y_true,average='weighted')
print("Precison_Score :	",pr,"\n Recall_Score :	",re,"\n F_Score : ",f1)


Precison_Score :	 0.6275555555555556 
 Recall_Score :	 0.37333333333333335 
 F_Score :  0.42760738942231047


In [14]:
y

array([[1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 1.

# Conclusion: 

### (1) We constructed an ANN with 4 hidden layers with 8,16,8 neurons for Egyptian Skull data
### (2)  The results of the classification metrics is 
#### Precison_Score :	 0.6275555555555556 
#### Recall_Score :	 0.37333333333333335 
#### F_Score :  0.42760738942231047
### (3) Thus, we illustrated this model with a sample prediction